src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupFreePeriodService.java

```diff
@@ -35,28 +46,108 @@ public Optional<TutorialGroupFreePeriod> findOverlappingPeriod(Course course, Tu
+
+        sessionsToCancel.forEach(session -> {
             session.setStatus(TutorialGroupSessionStatus.CANCELLED);
             // we set the status explanation to null, because the reason is now contained in the free period
             session.setStatusExplanation(null);
             session.setTutorialGroupFreePeriod(tutorialGroupFreePeriod);
         });
-        tutorialGroupSessionRepository.saveAll(overlappingSessions);
+        tutorialGroupSessionRepository.saveAll(sessionsToCancel);
     }
 
     /**
-     * Activate all tutorial group sessions that were cancelled because of the given free period
+     * Update all tutorial group sessions that were cancelled because of the given free period. Sessions that overlap with another free period are not activated.
      *
+     * @param course                  the course in which the free period is defined
      * @param tutorialGroupFreePeriod the free period
+     * @param onDeletion              true if the free period is deleted, false otherwise
+     */
+    public void updateOverlappingSessions(Course course, TutorialGroupFreePeriod tutorialGroupFreePeriod, boolean onDeletion) {
+        Set<TutorialGroupSession> overlappingSessions = tutorialGroupSessionRepository.findAllBetween(course, tutorialGroupFreePeriod.getStart(), tutorialGroupFreePeriod.getEnd());
+
+        findAndUpdateStillCanceledSessions(course, tutorialGroupFreePeriod, overlappingSessions, onDeletion);
+        findAndUpdateSessionsToReactivate(course, tutorialGroupFreePeriod, overlappingSessions);
+    }
+
+    /**
+     * Find and update tutorial group sessions that are still cancelled due to overlapping with another free period.
+     * If the Period that provided the reason initially is still present, the reason remains the same.
+     * If it is not, it finds the first overlapping free period and updates the session status, status explanation, and the tutorial group free period.
+     *
+     * @param course                  the course in which the free period is defined.
+     * @param tutorialGroupFreePeriod the free period.
+     * @param overlappingSessions     the set of sessions that overlap with the given free period.
+     * @param onDeletion              true if the free period is deleted, false otherwise.
      */
-    public void activateOverlappingSessions(TutorialGroupFreePeriod tutorialGroupFreePeriod) {
-        var overlappingSessions = tutorialGroupSessionRepository.findAllByTutorialGroupFreePeriodId(tutorialGroupFreePeriod.getId());
-        overlappingSessions.forEach(session -> {
+    private void findAndUpdateStillCanceledSessions(Course course, TutorialGroupFreePeriod tutorialGroupFreePeriod, Set<TutorialGroupSession> overlappingSessions,
+            boolean onDeletion) {
+        // find those sessions that are still overlapping with another free period
+        List<TutorialGroupSession> sessionsStillOverlappingWithFreePeriods = overlappingSessions.stream().filter(session -> {
```

Comment:
Why do you store this as a list when you get the sessions as a Set in the parameters?